wasm-chain-integration/src/v1/trie/low_level.rs

```diff
@@ -468,86 +467,295 @@ impl<V> CachedRef<V> {
+    /// Construct an iterator from the entire input.
+    pub fn new(data: &'a [u8]) -> Self {
+        Self {
+            data,
+            pos: 0,
+            len: 2 * data.len(),
+        }
+    }
+
+    /// Return the next chunk.
+    pub fn next(&mut self) -> Option<Chunk<4>> {
+        if self.pos < self.len {
+            // This access is safe since we have already checked the bound.
+            // Using an unchecked access here is about 10% faster for lookups on
+            // current benchmarks (tree size 100k).
+            let v = unsafe { *self.data.get_unchecked(self.pos / 2) };
+            if self.pos % 2 == 0 {
+                self.pos += 1;
+                Some(Chunk::new((v & 0xf0) >> 4))
+            } else {
+                self.pos += 1;
+                Some(Chunk::new(v & 0x0f))
+            }
+        } else {
+            None
+        }
+    }
 
-impl AsRef<[u8]> for Stem {
     #[inline(always)]
-    fn as_ref(&self) -> &[u8] {
-        match self {
-            Stem::Short(s) => {
-                let len = usize::from(s[0]);
-                &s[1..1 + len]
+    /// Convert the remaining part of the iterator into a stem.
+    pub fn to_stem(&self) -> Stem { self.last_to_stem(self.pos) }
+
+    // NB: Consumed - 1 to stem!
+    #[inline(always)]
+    /// Convert the consumed parts (**without the last element**) of the
+    /// iterator into a stem. The strange behaviour is what is needed.
+    pub fn consumed_to_stem(&self) -> Stem {
+        if self.pos == 0 {
+            Stem::empty()
+        } else {
+            let new_len = self.pos - 1;
+            if new_len % 2 == 0 {
+                Stem::new(self.data[..new_len / 2].into(), new_len)
+            } else {
+                let mut data = self.data[..new_len / 2].to_vec();
+                data.push(self.data[new_len / 2] & 0xf0);
+                Stem::new(data.into_boxed_slice(), new_len)
             }
-            Stem::Long(arc) => arc.as_ref(),
+        }
+    }
+
+    #[inline(always)]
+    /// Convert the data from the given position (inclusive) into a stem.
+    fn last_to_stem(&self, pos: usize) -> Stem {
+        let new_len = self.len - pos;
+        if pos % 2 == 0 {
+            let data = &self.data[pos / 2..];
+            Stem::new(data.into(), new_len)
+        } else {
+            let mut data = Vec::with_capacity(new_len + 1);
+            let mut left = (self.data[pos / 2] & 0x0f) << 4;
+            for &byte in &self.data[pos / 2 + 1..] {
+                data.push(left | (byte & 0xf0) >> 4);
+                left = (byte & 0x0f) << 4;
+            }
+            if new_len % 2 == 1 {
+                data.push(left);
+            }
+            Stem::new(data.into_boxed_slice(), new_len)
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[repr(transparent)]
+/// A wrapper around u8 that indicates an N-bit value. N must be between 1 and
+/// 8.
+struct Chunk<const N: usize> {
```

Comment:
Instead of having Chunk<4> everywhere would it perhaps be better to have a concrete type for this? When reading through the functions make use of the stem, it is rare you actually care about this - except for when performing binary search on the children. (<< 60) - which I believe could also be factored out. 

It may even perform better? I am not aware of the cost of 'riding' generics.

Reply:
This struct is `#[repr(transparent)`. There is no runtime cost.

Reply:
And in fact quite often you care about this. For instance the `.index` and `key` methods on the `KeyIndexPair` make use of the `N`.